File: CSharpCodeBase/utils/hump/vector2.cs
using System;
 
 using System.Collections.Generic;
 
 using System.Linq;
 
 using System.Text;
 
 using UnityEngine;
 
 using System.Collections;
 
 using MainGame.core;
 
 using UnityEngine.EventSystems;
 
 namespace MainGame{
 public class vector2 {
 public void init(self, x, y);
     this.x, this.y = x  ||  0, y  ||  0;
 }})
 public void __add( rhs ){
     ( this.x + rhs.x, this.y + rhs.y );
 }
 public void __sub( rhs ){
     ( this.x - rhs.x, this.y - rhs.y );
 }
 public void __mul( rhs ){
     ( this.x * rhs, this.y * rhs );
 }
 public void __div( rhs ){
     ( this.x / rhs, this.y / rhs );
 }
 public void Dot( rhs ){
     return this.x * rhs.x + this.y * rhs.y;
 }
 public void Cross( rhs ){
     ( this.x * rhs.y, -this.y * rhs.x );
 }
 public void __tostring(){
     return string.format("(%2.2f, %2.2f)", this.x, this.y);
 }
 public void __eq( rhs ){
     return this.x == rhs.x  &&  this.y == rhs.y;
 }
 public void DistSq(other){
     return (this.x - other.x)*(this.x - other.x) + (this.y - other.y)*(this.y - other.y);
 }
 public void Dist(other){
     return math.sqrt(self:DistSq(other));
 }
 public void LengthSq(){
     return this.x*this.x + this.y*this.y;
 }
 public void Length(){
     return math.sqrt(self:LengthSq());
 }
 public void Normalize(){
     var len = self:Length();
     if(len > 0  ){ 
         this.x = this.x / len;
         this.y = this.y / len;
     }
     return self;
 }
 public void GetNormalized(){
     return self / self:Length();
 }
 public void Get(){
     return this.x, this.y;
 }
 public void Vector2.Distance(v1, v2){
   return v1:Dist(v2);
 }
 }}

File: CSharpCodeBase/utils/hump/vector2.lua
local Class = require 'utils.hump.class'

local Vector2 = Class({init = function(self, x, y)
    self.x, self.y = x or 0, y or 0
end})

function Vector2:__add( rhs )
    return Vector2( self.x + rhs.x, self.y + rhs.y )
end

function Vector2:__sub( rhs )
    return Vector2( self.x - rhs.x, self.y - rhs.y )
end

function Vector2:__mul( rhs )
    return Vector2( self.x * rhs, self.y * rhs )
end

function Vector2:__div( rhs )
    return Vector2( self.x / rhs, self.y / rhs )
end

function Vector2:Dot( rhs )
    return self.x * rhs.x + self.y * rhs.y
end

function Vector2:Cross( rhs )
    return Vector2( self.x * rhs.y, -self.y * rhs.x )
end

function Vector2:__tostring()
    return string.format("(%2.2f, %2.2f)", self.x, self.y)
end

function Vector2:__eq( rhs )
    return self.x == rhs.x and self.y == rhs.y
end

function Vector2:DistSq(other)
    return (self.x - other.x)*(self.x - other.x) + (self.y - other.y)*(self.y - other.y)
end

function Vector2:Dist(other)
    return math.sqrt(self:DistSq(other))
end

function Vector2:LengthSq()
    return self.x*self.x + self.y*self.y
end

function Vector2:Length()
    return math.sqrt(self:LengthSq())
end

function Vector2:Normalize()
    local len = self:Length()
    if len > 0 then
        self.x = self.x / len
        self.y = self.y / len
    end
    return self
end

function Vector2:GetNormalized()
    return self / self:Length()
end

function Vector2:Get()
    return self.x, self.y
end

function Vector2.Distance(v1, v2)
  return v1:Dist(v2)
end

return Vector2